staging/src/k8s.io/api/storage/v1/types.go

```diff
@@ -560,3 +558,99 @@ type CSINodeList struct {
+
+// CSIStorageCapacity stores the result of one CSI GetCapacity call.
+// For a given StorageClass, this describes the available capacity in a
+// particular topology segment.  This can be used when considering where to
+// instantiate new PersistentVolumes.
+//
+// For example this can express things like:
+// - StorageClass "standard" has "1234 GiB" available in "topology.kubernetes.io/zone=us-east1"
+// - StorageClass "localssd" has "10 GiB" available in "kubernetes.io/hostname=knode-abc123"
+//
+// The following three cases all imply that no capacity is available for
+// a certain combination:
+// - no object exists with suitable topology and storage class name
+// - such an object exists, but the capacity is unset
+// - such an object exists, but the capacity is zero
+//
+// The producer of these objects can decide which approach is more suitable.
+//
+// They are consumed by the kube-scheduler if a CSI driver opts into
+// capacity-aware scheduling with CSIDriverSpec.StorageCapacity.
+type CSIStorageCapacity struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata. The name has no particular meaning. It must be
+	// be a DNS subdomain (dots allowed, 253 characters). To ensure that
+	// there are no conflicts with other CSI drivers on the cluster, the recommendation
+	// is to use csisc-<uuid>, a generated name, or a reverse-domain name which ends
+	// with the unique CSI driver name.
+	//
+	// Objects are namespaced.
+	//
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// NodeTopology defines which nodes have access to the storage
+	// for which capacity was reported. If not set, the storage is
+	// not accessible from any node in the cluster. If empty, the
+	// storage is accessible from all nodes. This field is
+	// immutable.
+	//
+	// +optional
+	NodeTopology *metav1.LabelSelector `json:"nodeTopology,omitempty" protobuf:"bytes,2,opt,name=nodeTopology"`
+
+	// The name of the StorageClass that the reported capacity applies to.
+	// It must meet the same requirements as the name of a StorageClass
+	// object (non-empty, DNS subdomain). If that object no longer exists,
+	// the CSIStorageCapacity object is obsolete and should be removed by its
+	// creator.
+	// This field is immutable.
+	StorageClassName string `json:"storageClassName" protobuf:"bytes,3,name=storageClassName"`
+
+	// Capacity is the value reported by the CSI driver in its GetCapacityResponse
+	// for a GetCapacityRequest with topology and parameters that match the
+	// previous fields.
+	//
+	// The semantic is currently (CSI spec 1.2) defined as:
+	// The available capacity, in bytes, of the storage that can be used
+	// to provision volumes. If not set, that information is currently
+	// unavailable and treated like zero capacity.
```

Comment:
Unless MaximumVolumeSize is defined?

Reply:
If MaximumVolumeSize is set and Capacity is not, then MaximumVolumeSize will be used.

This doesn't occur in practice right now. Because CSI drivers always report `available_capacity`.

Reply:
Document what is allowed in the comment, and update the validation accordingly.

Reply:
> Document what is allowed in the comment

The documentation is correct: both fields are optional and the creator decides what it provides.

What I can clarify is the interpretation of "MaximumVolumeSize non-nil and Capacity nil": as implemented and IMHO implied, the scheduler then uses "MaximumVolumeSize".

Reply:
@alculquicondor Let's discuss the revised comments in https://github.com/kubernetes/enhancements/pull/3233 first, see https://github.com/kubernetes/enhancements/pull/3233/commits/c1505fef482e0e38f78e3d4e0b58e751133b40bd.

Reply:
Then I'll copy them into the various types.go files and update the generated files.

Reply:
I've updated the comments as agreed in https://github.com/kubernetes/enhancements/pull/3233#discussion_r821853048.